File: LinkedList/Python/DoublyLinkedList.py
class Node:
    def __init__(self, element):
        self.element = element
        self.next = None
        self.prev = None


class DoublyLinkedList:
    def __init__(self):
        self.head = None

    def insert_head(self, val):
        if self.head is None:
            self.head = Node(val)
            return
        temp_node = Node(val)
        temp_node.next = self.head
        self.head.prev = temp_node
        self.head = temp_node

    def insert_tail(self, val):
        temp_node = Node(val)
        if self.head is None:
            self.head = temp_node
            return
        iterate = self.head
        while iterate.next:
            iterate = iterate.next
        iterate.next = temp_node
        temp_node.prev = iterate

    def insert_after(self, insert_after, val):  # IDE suggests method to be static since "self" isn't used...
        if insert_after is None:
            print("Node to insert new data after cannot be null.")
            return
        temp_node = Node(val)
        temp_node.next = insert_after.next
        insert_after.next = temp_node
        temp_node.prev = insert_after
        if temp_node.next:
            temp_node.next.prev = temp_node

    def delete_by_position(self, position):
        if self.head is None:
            print("Can't delete a non existent LL.")
            return
        iterate = self.head
        if position == 0:
            self.head = iterate.next
            self.head.prev = None
        for _ in range(position - 1):
            iterate = iterate.next
        if iterate.next is None or iterate is None:  # Since next is the node to be deleted
            print("Node isn't in LL.")
            return
        new_next = iterate.next.next
        iterate.next = new_next
        if iterate.next is not None:
            iterate.next.prev = iterate.prev

    def search(self, val):
        iterate = self.head
        while iterate is not None:
            if iterate.element == val:
                return True
            iterate = iterate.next
        return False

    def sort(self):  # Use Bubble sort
        if self.head is None or self.head.next is None:
            print("Nothing to sort...")
            return
        temp_head = self.head
        next_to_curr = Node(None)
        while temp_head is not None:
            next_to_curr = temp_head.next
            while next_to_curr is not None:
                if temp_head.element > next_to_curr.element:
                    temp_head.element, next_to_curr.element = next_to_curr.element, temp_head.element
                next_to_curr = next_to_curr.next
            temp_head = temp_head.next

    def print_ll(self):
        iterate = self.head
        while iterate is not None:
            print(str(iterate.element) + "   ", end='')
            iterate = iterate.next
        print()


my_ll = DoublyLinkedList()
my_ll.insert_head(5)
my_ll.insert_after(my_ll.head, 10)
my_ll.insert_head(15)
my_ll.insert_tail(20)
my_ll.insert_after(my_ll.head.next.next.next, 25)

my_ll.print_ll()

my_ll.delete_by_position(2)
my_ll.delete_by_position(2)
my_ll.insert_after(my_ll.head.next, 500)
my_ll.print_ll()

my_ll.sort()
my_ll.print_ll()

File: LinkedList/JAVA/src/implementing_linkedlist/DoublyLinkedList.java
package implementing_linkedlist;

public class DoublyLinkedList {
	Node head;

	class Node {
		int data;
		Node next;
		Node prev;

		Node(int data) {
			this.data = data;
			next = null;
			prev = null;
		}
	}

	public void insertHead(int newData) {
		if (head == null) {
			head = new Node(newData);
			return;
		}

		Node newHead = new Node(newData);
		newHead.next = head;
		if (head != null)
			head.prev = newHead;
		head = newHead;
	}

	public void insertTail(int newData) {
		Node newTail = new Node(newData);

		if (head == null) { // there's no head node, create it using the new data
			head = new Node(newData);
			return;
		}

		newTail.next = null;
		Node last = head; // iterate to the tail of the existing LL
		while (last.next != null)
			last = last.next;
		last.next = newTail;
		newTail.prev = last;
	}

	public void insertAfter(Node toInsertAfter, int newData) {
		if (toInsertAfter == null) {
			System.out.println("Node to insert new data after cannot be null.");
			return;
		}

		Node newNode = new Node(newData);
		newNode.next = toInsertAfter.next; // prevNode -> newNode -> nextNode
		toInsertAfter.next = newNode;
		newNode.prev = toInsertAfter;

		if (newNode.next != null)
			newNode.next.prev = newNode;
	}

	public void deleteByPosition(int position) {
		if (head == null) {
			System.out.println("Can't delete a non existant LL.");
			return;
		}

		Node iterHead = head;

		if (position == 0) {
			head = iterHead.next; // iterate to the next node, meaning to delete first (head) node
			head.prev = null;
			return;
		}

		for (int i = 0; i < position - 1 && iterHead != null; i++)
			iterHead = iterHead.next;

		if (iterHead == null || iterHead.next == null) {
			System.out.println("Node isn't in LL.");
			return;
		}

		Node temp = iterHead;

		// iterHead.next != null
		iterHead.next = temp.next.next;

		// now iterHead.next == previously iterHead.next.next
		if (iterHead.next != null)
			iterHead.next.prev = temp.prev;
	}

	public boolean search(Node head, int data) {
		Node iterHead = head;
		while (iterHead != null) {
			if (iterHead.data == data)
				return true;
			iterHead = iterHead.next;
		}
		return false;
	}

	// Use Merge-Sort:
	public Node getMid(Node head) {
		if (head == null)
			return head;

		Node slow = head;
		Node fast = head;

		while (fast.next != null && fast.next.next != null) {
			slow = slow.next;
			fast = fast.next.next;
		}
		return slow;
	}

	public Node sortMergeLL(Node head) {
		if (head == null || head.next == null)
			return head;

		Node mid = getMid(head);
		Node nextMid = mid.next;

		mid.next = null;

		Node left = sortMergeLL(head);
		Node right = sortMergeLL(nextMid);

		Node sortedLL = merge(left, right);
		return sortedLL;
	}

	public Node merge(Node a, Node b) { // merge left and right lists
		Node merged = null;

		if (a == null)
			return b;
		if (b == null)
			return a;

		if (a.data <= b.data) {
			merged = a;
			merged.next = merge(a.next, b);
		} else {
			merged = b;
			merged.next = merge(a, b.next);
		}

		return merged;
	}

	public void printLL() {
		Node iterHead = head;
		while (iterHead != null) {
			System.out.print(iterHead.data + "	");
			iterHead = iterHead.next;
		}
		System.out.println();
	}

	public static void main(String[] args) {
		DoublyLinkedList linkedList = new DoublyLinkedList();

		linkedList.insertHead(5);
		linkedList.insertAfter(linkedList.head, 1);
		linkedList.insertTail(0);
		linkedList.insertAfter(linkedList.head.next, 7);
		linkedList.insertAfter(linkedList.head.next.next, 10);

		linkedList.printLL();

		linkedList.deleteByPosition(1);

		linkedList.printLL();

		Node sortedHead = linkedList.sortMergeLL(linkedList.head);
		linkedList.head = sortedHead;
		linkedList.printLL();
	}
}